### PR TITLE
refactor(automation): move drive-green to post-loop terminal stage

### LINKED ISSUE #818: drive-green: move from per-loop phase to per-repo post-loop terminal stage
## Summary

`drive-green` is currently the third entry in `ALL_PHASES` (`hephaestus/automation/loop_runner.py:77-81`) and is iterated inside the per-loop, per-repo `for phase in ALL_PHASES` loop. It is then carved out with a special-case skip at `loop_runner.py:918-924` that only lets it run when `loop_idx == cfg.loops`. That model is wrong: `drive-green` is a **post-loop terminal stage** that should run once per repo *after* all loop iterations of `plan`+`implement` are finished.

## Why it's wrong (concrete consequence)

Combined with the early-exit logic at `loop_runner.py:1108-1121`, `--phases drive-green --loops N` for any `N > 1` is unreachable code: loops 1..N-1 trivially produce zero work AND skip `drive-green`, so early-exit fires before the final loop is reached.

Repro (2026-05-30):

```
$ pixi run hephaestus-automation-loop --phases drive-green --repos ProjectHephaestus,ProjectMnemosyne --loops 5
[ProjectHephaestus] phase drive-green SKIP (not final loop)
[ProjectMnemosyne] phase drive-green SKIP (not final loop)
Early exit after loop 1/5: full pass produced 0 new plans and 0 non-skipped reviews
✓ Completed 1 of 5 loop(s)
```

## Required changes

- Shrink `ALL_PHASES` to `(\"plan\", \"implement\")` (`loop_runner.py:77-81`).
- Add `ALL_POST_LOOP_STAGES = (\"drive-green\",)`.
- Delete the `loop_runner.py:918-924` \"not final loop\" special case — there's no longer a phase to gate.
- Refactor `run_loop` (`loop_runner.py:1050+`) so it runs the per-loop iteration body to completion (or early-exit), then iterates post-loop stages **once per repo**:

  ```python
  for loop_idx in 1..cfg.loops:
      for repo in repos:
          for phase in ALL_PHASES:                # plan, implement only
              ...
      if early_exit_condition(...):
          break
  # Post-loop, runs regardless of early-exit
  for repo in repos:
      for stage in ALL_POST_LOOP_STAGES:
          if stage in cfg.phases:                 # or cfg.stages
              run_stage(repo, stage, ...)
  ```

- Update the predecessor-ordering warning at `loop_runner.py:402-405` — \"`drive-green` without `implement`\" becomes a legitimate operator intent (\"drive existing PRs without opening new work\"), not a footgun.
- Update the module docstring at `loop_runner.py:1-19` and the `ALL_PHASES` comment at `loop_runner.py:72-76` to describe `plan → implement` as the loop body and `drive-green` as a per-repo post-loop terminal stage.
- Update `_summarize_loop` / `RepoResult` to surface post-loop stage outcomes distinctly from per-loop phase outcomes.

## Acceptance criteria

- `pixi run hephaestus-automation-loop --phases drive-green --loops 5` runs `drive-green` **once per repo** after the loop body (today: 0 times).
- `pixi run hephaestus-automation-loop --loops 5` (default phase set) runs `plan` + `implement` up to 5 times each, then `drive-green` **once per repo**, including when early-exit cuts the loop body short on loop 1.
- `pixi run hephaestus-automation-loop --phases plan,implement --loops 5` does NOT invoke `drive-green` even after the loop completes.
- New unit tests in `tests/unit/automation/`:
  - `drive-green` selected alone → runs once per repo as post-loop, regardless of `--loops` value.
  - `drive-green` selected alongside loop phases → runs ONCE after iteration finishes, not once per iteration.
  - Early-exit on loop 1 still triggers the post-loop `drive-green` for each repo.
  - `--phases plan` (no drive-green) → no `drive-green` invocation anywhere.

## Out of scope (filed separately)

- `drive-green` discovers work by open PRs, not open issues — see follow-up issue.
- `--issues` is `required=True` in `scripts/drive_prs_green.py` — see follow-up issue.
- `_list_open_issue_numbers` is `@me`-scoped — see follow-up issue.

## Related

- 3-stage collapse came in #677 / #679 — kept `drive-green` in `ALL_PHASES` for migration simplicity; this issue takes the next step.
- Original combined report: #817 (this issue and the three siblings supersede it; #817 can be closed when all four ship).

### PR DESCRIPTION
**Pull request by mvillmow**
## Summary

Refactor `loop_runner.py` so `drive-green` transitions from "third loop phase, final-loop only" to a dedicated post-loop terminal stage that runs once per repo *after* all loop iterations (including early-exit). This fixes the unreachable code issue: `--phases drive-green --loops N` now correctly runs `drive-green` once per repo regardless of N, with no spurious "not final loop" skips on loops 1..N-1.

## Test Plan

- [x] All 142 tests pass (89 existing + 39 early-exit + 14 new post-loop)
- [x] mypy: no errors
- [x] ruff: all checks passed
- [x] Acceptance criterion 1: `--phases drive-green --loops 5` runs drive-green once per repo
- [x] Acceptance criterion 2: `--loops 5` default runs plan+implement loop, then drive-green once per repo (including on early-exit)
- [x] Acceptance criterion 3: `--phases plan,implement --loops 5` does NOT invoke drive-green
- [x] Acceptance criterion 4: Early-exit on loop 1 still triggers post-loop drive-green per repo

🤖 Generated with [Claude Code](https://claude.com/claude-code)

Closes #818

### CODE REVIEW

**Review by mvillmow — COMMENTED** [review]
POLICY VIOLATION: Closes, auto-merge-enabled, signed-commits — PR body is empty (no Closes #818), auto_merge_enabled=true must be false until review GO, commit signing state array is empty (no verified signatures).

File: hephaestus/automation/loop_runner.py [hunk 1]
```diff
@@ -72,18 +77,29 @@ def _default_phase_timeout_s() -> float:
...
+# new arg); the runner partitions selected names into loop phases vs.
+# post-loop stages internally.
+ALL_POST_LOOP_STAGES: tuple[str, ...] = (
     "drive-green",
 )
 
-FINAL_LOOP_ONLY_PHASES: frozenset[str] = frozenset({"drive-green"})
+# Union used by --phases validation. Operators may select any combination
+# of loop phases and post-loop stages on the same flag.
+ALL_SELECTABLE: tuple[str, ...] = ALL_PHASES + ALL_POST_LOOP_STAGES
```
Comment (mvillmow) [inline hephaestus/automation/loop_runner.py:103]: Now that `ALL_PHASES` shrank to `("plan", "implement")`, the `LoopConfig.phases` dataclass default (line 304, `phases: tuple[str, ...] = ALL_PHASES`) diverges from the CLI `--phases` default (`ALL_SELECTABLE`). Pre-#818 a programmatic `LoopConfig()` included drive-green; now it silently excludes it. `main()` always passes phases explicitly so the CLI is unaffected, but either point `LoopConfig.phases` at `ALL_SELECTABLE` to restore default parity, or add a comment on the field documenting that the dataclass default is loop-body-only by design.

File: tests/unit/automation/test_loop_runner_early_exit.py [hunk 2]
```diff
@@ -448,18 +449,32 @@ def fake_process(repo: str, loop_idx: int, cfg: LoopConfig) -> RepoResult:
...
+        promoted to a post-loop terminal stage, early-exit fires after loop 1
+        on a zero-work pass — drive-green is invoked AFTER the iteration body
+        by ``_run_post_loop_stages`` regardless. The loop_runner_post_loop.py
+        suite asserts the post-loop invocation; this test pins the early-exit
+        behaviour for the loop body itself.
+        """
         projects = tmp_path
         (projects / "r1" / ".git").mkdir(parents=True)
-        cfg = LoopConfig(loops=3, projects_dir=projects)
+        cfg = LoopConfig(loops=3, projects_dir=projects, org="testorg")
```
Comment (mvillmow) [inline tests/unit/automation/test_loop_runner_early_exit.py]: This test is named `..._with_post_loop_drive_green_selected`, but `LoopConfig(loops=3, projects_dir=projects, org="testorg")` uses the default `phases=ALL_PHASES`, which post-#818 is `("plan", "implement")` — drive-green is NOT selected, contradicting the name and docstring and making the test nearly identical to `test_early_exit_fires_on_zero_work_pass`. Pass `phases=("plan", "implement", "drive-green")` (the docstring's claimed scenario) so the test pins early-exit behavior with drive-green actually selected.

**Review by mvillmow — COMMENTED** [review]
GO: all 4 acceptance criteria implemented and tested; dead symbols verified removed; post-loop runs on both exit paths. Two minor findings: LoopConfig default-phases divergence; one mislabeled test.

**Review by mvillmow — COMMENTED** [review]
GO: rebase onto #819 absorbed correctly; gates migrated faithfully to _run_post_loop_stages; all acceptance criteria tested; dead symbols gone. Two minor findings: LoopConfig default divergence (now bites two tests); mislabeled early-exit test.

File: hephaestus/automation/loop_runner.py [hunk 3]
```diff
@@ -73,18 +78,27 @@ def _default_phase_timeout_s() -> float:
...
+# finish (exhaustion or early-exit). Per #818, drive-green belongs here
+# because it polls existing PRs — it is a terminal action, not an iteration
+# step. Operators select it with ``--phases drive-green`` (same flag, no
+# new arg); the runner partitions selected names into loop phases vs.
+# post-loop stages internally.
+ALL_POST_LOOP_STAGES: tuple[str, ...] = ("drive-green",)
+
+# Union used by --phases validation. Operators may select any combination
+# of loop phases and post-loop stages on the same flag.
+ALL_SELECTABLE: tuple[str, ...] = ALL_PHASES + ALL_POST_LOOP_STAGES
```
Comment (mvillmow) [inline hephaestus/automation/loop_runner.py:103]: Now that `ALL_PHASES` shrank to `("plan", "implement")`, the `LoopConfig.phases` dataclass default (line 299) diverges from the CLI `--phases` default (`ALL_SELECTABLE`). This already bites inside this PR: `test_early_exit_when_drive_green_selected_but_no_failing_prs` (test_loop_runner_early_exit.py:498) still says 'default phases incl. drive-green' — false post-#818 — and its `_count_failing_prs` patch is vestigial. Either point the dataclass default at `ALL_SELECTABLE` (and add the `_run_post_loop_stages` patch to that test, or its unpatched post-loop block will run a real `_rebase_main` and break its `max(loop_idx) == 1` assertion), or document on the field that the dataclass default is loop-body-only by design and fix that test's stale comment.

File: tests/unit/automation/test_loop_runner_early_exit.py [hunk 4]
```diff
@@ -448,22 +450,32 @@ def fake_process(repo: str, loop_idx: int, cfg: LoopConfig) -> RepoResult:
...
+        promoted to a post-loop terminal stage, early-exit fires after loop 1
+        on a zero-work pass — drive-green is invoked AFTER the iteration body
+        by ``_run_post_loop_stages`` regardless. The loop_runner_post_loop.py
+        suite asserts the post-loop invocation; this test pins the early-exit
+        behaviour for the loop body itself.
         """
         projects = tmp_path
         (projects / "r1" / ".git").mkdir(parents=True)
-        cfg = LoopConfig(loops=3, projects_dir=projects)  # default phases incl. drive-green
+        cfg = LoopConfig(loops=3, projects_dir=projects, org="testorg")
```
Comment (mvillmow) [inline tests/unit/automation/test_loop_runner_early_exit.py]: This test is named `..._with_post_loop_drive_green_selected` and its closing comment says 'even with drive-green selected', but `LoopConfig(loops=3, projects_dir=projects, org="testorg")` uses the default `phases=ALL_PHASES`, which post-#818 is `("plan", "implement")` — drive-green is NOT selected, making this nearly a duplicate of `test_early_exit_fires_on_zero_work_pass`. Pass `phases=("plan", "implement", "drive-green")`; the existing `_run_post_loop_stages` patch keeps it hermetic while actually testing the claimed scenario.

File: hephaestus/automation/loop_runner.py [hunk 5]
```diff
@@ -1405,14 +1484,26 @@ def run_loop(cfg: LoopConfig, repos: list[str]) -> list[RepoResult]:
...
+    # iteration body exited (exhaustion or early-exit). Per #818.
+    post_loop_results = _run_post_loop_stages(cfg, repos)
+    all_results.extend(post_loop_results)
+
+    # Report actual loops run (may be less than cfg.loops due to early
+    # exit). Post-loop RepoResults always have non-empty post_loop_phases
+    # AND empty phases, so filtering by ``not r.post_loop_phases`` reliably
+    # excludes them — this prevents the post-loop record's loop_idx=cfg.loops
+    # from spuriously inflating the count when early-exit cut the loop short.
+    per_loop_results = [r for r in all_results if not r.post_loop_phases]
```
Comment (mvillmow) [inline hephaestus/automation/loop_runner.py]: The `not r.post_loop_phases` filter mis-classifies crashed/uncloned post-loop RepoResults as per-loop results. In `_run_post_loop_stages`, a repo whose `.git` is missing (line 1370-1373) or whose `_rebase_main`/`_list_open_issue_numbers` raises before any stage runs (line 1405-1408) produces a RepoResult with BOTH `phases` AND `post_loop_phases` empty. Such a record (with `loop_idx=cfg.loops`) passes the `not r.post_loop_phases` test and lands in `per_loop_results`, so `max(...)` reports `actual_loops=cfg.loops` even when early-exit cut the loop body short at loop 1 — directly violating the inline comment's asserted invariant ('Post-loop RepoResults always have non-empty post_loop_phases'). Impact is limited to the cosmetic '✓ Completed N of M loop(s)' line (no control-flow effect), but the fix is cheap: tag post-loop records explicitly (e.g. a `is_post_loop: bool` field on RepoResult, set in `_run_post_loop_stages`) and filter on that instead of inferring intent from list emptiness.

**Review by mvillmow — COMMENTED** [review]
drive-green correctly moved to post-loop terminal stage; all 4 acceptance criteria tested; unreachability fixed. One minor actual_loops reporting edge case on post-loop crash.

File: tests/unit/automation/test_loop_runner_early_exit.py (hunk 4 above)
Comment (mvillmow) [reply tests/unit/automation/test_loop_runner_early_exit.py]: Addressed in 0dc741a4 (test_loop_runner_early_exit.py:482): the test now passes explicit phases=('plan','implement','drive-green') so it actually selects drive-green, matching its name/docstring; the _run_post_loop_stages stub keeps it hermetic.

File: hephaestus/automation/loop_runner.py (hunk 3 above)
Comment (mvillmow) [reply hephaestus/automation/loop_runner.py:103]: Addressed in 0dc741a4: documented on the field (loop_runner.py:307) that the LoopConfig.phases dataclass default is loop-body-only (ALL_PHASES) by design, contrasting it with the CLI --phases default (ALL_SELECTABLE). The stale 'default phases incl. drive-green' comment is fixed and the vestigial _count_failing_prs patch was swapped for a _run_post_loop_stages stub.

File: hephaestus/automation/loop_runner.py (hunk 5 above)
Comment (mvillmow) [reply hephaestus/automation/loop_runner.py]: Addressed in 0dc741a4: added an explicit is_post_loop bool on RepoResult (loop_runner.py:229), set True in _run_post_loop_stages (1367), and filtered per_loop_results on it (1497) instead of inferring from list emptiness — so crashed/uncloned post-loop records with both phase lists empty no longer leak into per_loop_results.
